File: SOLID Principles/C#/ISP.cs
using System;

namespace ISP
{
  // BAD IMPLEMENTATION
  interface IPrinter
  {
    void Print();
    void Scan();
    void Fax();
  }

  class MultiFunctionPrinter : IPrinter
  {
    public void Print() { Console.WriteLine("Printing document"); }
    public void Scan() { Console.WriteLine("Scanning document"); }
    public void Fax() { Console.WriteLine("Faxing document"); }
  }

  class SimplePrinter : IPrinter
  {
    public void Print() { Console.WriteLine("Printing document"); }

    public void Scan()
    {
      throw new NotImplementedException("SimplePrinter cannot scan");
    }

    public void Fax()
    {
      throw new NotImplementedException("SimplePrinter cannot fax");
    }
  }

  // Actual IMPLEMENTATION (ISP Applied)

  interface IAPrinter
  {
    void Print();
  }

  interface IScanner
  {
    void Scan();
  }

  interface IFax
  {
    void Fax();
  }


  interface IMultiPrinter : IAPrinter, IScanner, IFax { }

  public class MultiFuncPrinter : IMultiPrinter
  {
    public void Print()
    {
      Console.WriteLine("Printing document");
    }

    public void Scan()
    {
      Console.WriteLine("Scanning document");
    }

    public void Fax()
    {
      Console.WriteLine("Faxing document");
    }
  }

  public class ASimplePrinter : IAPrinter
  {
    public void Print()
    {
      Console.WriteLine("Printing document");
    }
  }

  class Program
  {
    static void Main(string[] args)
    {
      IAPrinter simple = new ASimplePrinter();
      simple.Print();

      IMultiPrinter multi = new MultiFuncPrinter();
      multi.Fax();
    }
  }
}

File: Typescript/SOLID Principles/ISP.ts
// I - Interface Segregation Principle (ISP)

// Bad Implementation Example:
interface IPrinter {
  print(): void;
  scan(): void;
  fax(): void;
}

class MultiFunctionPrinter implements IPrinter {
  print() { console.log("Printing document") }
  scan() { console.log("Scanning document") }
  fax() { console.log("Faxing document") }
}

class SimplePrinter implements IPrinter {
  print() { console.log("Printing document") }
  scan() { throw new Error("SimplePrinter cannot scan") }
  fax() { throw new Error("SimplePrinter cannot fax") }
}



// Actual Implementation
interface IAPrinter {
  print(): void;
}

interface IScanner {
  scan(): void;
}

class AMultiPrinter implements IAPrinter, IScanner {
  print() { console.log("Printing document") }
  scan() { console.log("Scanning document") }
}

class ASimplePrinter implements IAPrinter {
  print() { console.log("Printing document") }
}



